### DISCUSSION TITLE
Where do Extensions go in a Python3 install?

### DISCUSSION THREAD

**Discussion by raowenqmul** [DIRACGrid/DIRAC#5943] [Q&A] [answered]
Hello, 
GridPP has a Dirac Extension called GridPPDIRAC.

When I installed DIRAC (v7r3p12) on python2 with our GridPP Dirac Extension at /opt/DIRAC/GridPPDIRAC
and with an install.cfg containing the following lines:
```
## GridPPDIRAC
  GridPP = yes
  #Extensions += GridPP
  ExtraModules += GridPPDIRAC
  Agents += Configuration/UsersAndGroupsAgent
  Agents += Configuration/AutoBdii2CSAgent
```
Then the installer auto discovers the 2 agents from /opt/DIRAC/GridPPDIRAC as can be seen from this extract of the installer output:
```
Trying to autodiscover Configuration/AutoBdii2CSAgent
Trying to load GridPPDIRAC.ConfigurationSystem.Agent.AutoBdii2CSAgent
Called from module GridPPDIRAC.ConfigurationSystem.private.AutoBDIISEs
Found GridPPDIRAC.ConfigurationSystem.Agent.AutoBdii2CSAgent
```

However when I installed DIRAC (7.3.12) on python3  and ask for installation in /opt/DIRAC with
`./install_site.sh -i /opt/dirac -v install.cfg
`
and with an install.cfg containing the same lines as above then the auto discovery of agents in  /opt/DIRAC/GridPPDIRAC fails with the following installer output:

```
2022-03-10 14:46:49 UTC dirac-setup-site [140179810096960] INFO: Trying to autodiscover Configuration/AutoBdii2CSAgent
2022-03-10 14:46:49 UTC dirac-setup-site [140179810096960] VERBOSE: Trying to load WebAppDIRAC.ConfigurationSystem.Agent.AutoBdii2CSAgent
2022-03-10 14:46:49 UTC dirac-setup-site [140179810096960] VERBOSE: Trying to load DIRAC.ConfigurationSystem.Agent.AutoBdii2CSAgent
2022-03-10 14:46:49 UTC dirac-setup-site [140179810096960] VERBOSE: Trying to load DIRACWebAppResources.ConfigurationSystem.Agent.AutoBdii2CSAgent
2022-03-10 14:46:49 UTC dirac-setup-site [140179810096960] ERROR: Software for agent Configuration/AutoBdii2CSAgent is not installed
```

My question is how do I tell DIRAC on python3 where to look in the filesystem for our GridPP extension?

I may well have other porting work to do but I think getting DIRAC to know where to look it probably the first step so any pointers would be gratefully received!

Many thanks for any ideas!

Alex Owen

**Comment by aldbr** [comment, marked as answer]
Python3 installation is a bit different from Python2 installation.
Here is the `Server installation` documentation: https://dirac.readthedocs.io/en/latest/AdministratorGuide/ServerInstallations/InstallingDiracServer.html

In the [Primary server installation](https://dirac.readthedocs.io/en/latest/AdministratorGuide/ServerInstallations/InstallingDiracServer.html#primary-server-installation) section, you can see differences between Python2 and Python3 installations:
- location, version, extension and additional packages related to the installation are passed via the command line interface in Python3 (while in Python2, you had to write them in `install.cfg` as you said).

> Run install_site.sh giving the edited configuration file as the argument. The configuration file must have .cfg extension (CFG file). While not strictly necessary, it’s advised that a version is added with the ‘-v’ switch (pick the most recent one, see here<https://pypi.org/project/DIRAC/#history>). In the same way, extensions have to be added with the ‘-e’ switch (the name of the extension should be complete). Finally, further pip packages (e.g. WebAppDIRAC) can follow with the ‘-p’ switch, which can be repeated multiple times:
> ```bash
> ./install_site.sh -i /opt/dirac [-v <x.y.z>] [-e <extension>] [-p <extra-pip-install>] install.cfg
> ```
